### PR TITLE
Orca unique() on DataShards on Spark

### PR DESCRIPTION
**Pull request by jenniew**
Implementation of unique() on DataShards of Pandas Series and Index.
Also implement select operation with [col] because we need to do unique() on df[col].
May add support for DataShards of numpy later. 
Related issue: https://github.com/analytics-zoo/orca/issues/16

### CODE REVIEW

**Comment by jason-dai** [comment]
There are two issues that need to be figured out:
1. For Spark based implementation, we need to automatically manage data cache
   ```
   shards = read_csv(..) #need to be automatically cached
   loc_df = shards[location] 
   location = loc_df.unique()
   shards.transform_shard(...) #need to cache the new shards, and delete old ones
   ```
2. For Ray based implementation, we need to manage the backing memory space 
   ```
   shards = read_csv(..) #df stored in Actor
   loc_df = shards[location] #additional df stored in Actor - can consume a lot of memory
                             #only store index - loc_df and shards share the same data
   location = loc_df.unique()
   shards.transform_shard(...) #update to shards also changes loc_df
   ```

**Comment by jenniew** [comment]
Is there any possibility that user do not do aggregation on loc_df,  or they may do loc_df.transform_shard()?

**Comment by jenniew** [comment]
Updated to support shards.unique('location'). Currently only support select column name to do unique operation in DataFrame.

File: pyzoo/zoo/orca/data/shard.py [hunk 1]
```diff
@@ -151,3 +151,23 @@ def merge(iterator):
...
                             " of Pandas DataFrame")
+
+    def unique(self, key):
+        import pandas as pd
+        elem_class, columns = self.rdd.map(
+            lambda data: (type(data), data.columns) if isinstance(data, pd.DataFrame)
+            else (type(data), None)).first()
+        if issubclass(elem_class, pd.DataFrame):
+            import numpy as np
+            if key is None:
```
Comment (jason-dai) [inline pyzoo/zoo/orca/data/shard.py:162]: why can't key be None?

File: pyzoo/zoo/orca/data/shard.py [hunk 2]
```diff
@@ -151,3 +151,23 @@ def merge(iterator):
...
+                raise Exception("Cannot apply unique operation on Datashards of Pandas Dataframe"
+                                " without column name")
+            if key in columns:
+                rdd = self.rdd.map(lambda df: df[key].unique().tolist())
+                result = rdd.reduce(lambda list1, list2: np.array(list(set(list1 + list2))))
```
Comment (jason-dai) [inline pyzoo/zoo/orca/data/shard.py:167]: need to perform unique operation in reduce, no?

File: pyzoo/zoo/orca/data/shard.py (hunk 1 above)
Comment (jenniew) [reply pyzoo/zoo/orca/data/shard.py:162]: if key is None, we may want to apply unique() on DataFrame. But DataFrame has no unique() operation. unique() works on 1-d array like

File: pyzoo/zoo/orca/data/shard.py (hunk 2 above)
Comment (jenniew) [reply pyzoo/zoo/orca/data/shard.py:167]: set(list1+list2) is to do unique. Or we can also run pd.unique(list1 + list2) in reduce, too.

Comment (jenniew) [reply pyzoo/zoo/orca/data/shard.py]: change to pd.unique(list1 + list2) which is faster.

**Comment by jason-dai** [comment]
Is this PR ready for merge?

**Comment by jenniew** [comment]
> Is this PR ready for merge?

Jenkins server always gets some problems in recently days. Will check with Dongjie to see if I can merge.

**Pull request merged**